ui/src/base/source_map_utils.ts

```diff
@@ -0,0 +1,344 @@
+// Copyright (C) 2025 The Android Open Source Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+export interface SourceMap {
+  version: number;
+  sources: string[];
+  names: string[];
+  mappings: string;
+  file?: string;
+}
+
+export interface MappingSegment {
+  genCol: number;
+  sourceIndex: number;
+  sourceLine: number;
+  sourceCol: number;
+  nameIndex: number;
+  hasName: boolean;
+}
+
+export interface ProcessedSourceMap {
+  sourceMap: SourceMap;
+  // Array of segments for each line, sorted by genCol for binary search
+  lineSegments: MappingSegment[][];
+}
+
+export interface MappedPosition {
+  source: string | null;
+  line: number | null;
+  column: number | null;
+  name: string | null;
+}
+
+// VLQ (Variable Length Quantity) base64 decoder for source maps
+const BASE64_CHARS =
+  'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
+
+const VLQ_BASE_SHIFT = 5;
+const VLQ_BASE = 1 << VLQ_BASE_SHIFT; // 32
+const VLQ_BASE_MASK = VLQ_BASE - 1; // 31
+const VLQ_CONTINUATION_BIT = VLQ_BASE; // 32
+
+const sourceMapCache = new Map<string, ProcessedSourceMap>();
+
+// Extend the global interface to include our custom property
+interface PerfettoGlobal {
+  __SOURCEMAPS?: Record<string, SourceMap>;
+}
```

Comment:
I think you can just do

declare global {
  var __SOURCEMAPS ?: Record<string, SourceMap>;
}

and this avoid the need of "self as unknown as Perfettoglobal"... as you can just do self.__SOURCEMAPS

Reply:
The problem is these declares are leaky - all typescript files will be able to use __SOURCEMAPS, but we have no reason to use it in any file other than this one.

For some reason this seemed like a terrible idea yesterday, but thinking about it now though maybe it's not such a big deal....